Warn about tech preview of checksums.

jessfraz opened the pull request:
ping @icecrime @crosbymichael @shykes @tianon

jessfraz committed:
Warn about tech preview of checksums.

Docker-DCO-1.1-Signed-off-by: Jessica Frazelle <jess@docker.com> (github: jfrazelle)

tiborvass commented on graph/pull.go:
```diff
@@ -491,7 +491,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 				io.Copy(tmpFile, utils.ProgressReader(ioutil.NopCloser(tarSumReader), int(l), out, sf, false, utils.TruncateID(img.ID), "Downloading"))
 
-				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum", nil))
+				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum, this is a tech preview, don't rely on it for security yet.", nil))
```
maybe replace the first comma `,` by ` - ` ? other than that, looks good!

jfrazelle replied:
cool will do!

jfrazelle commented on graph/pull.go:
```diff
@@ -428,7 +428,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 
 	if verified {
-		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified"))
+		out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified.\nThis is a tech preview, don't rely on it for security yet."))
```
I actually think we should just remove this one, it is only verifying the manifest not the layers.

tianon replied:
+1

icecrime commented on graph/pull.go:
```diff
@@ -428,10 +428,10 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
-		out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
+		log.Printf("Image manifest verified for %s:%s", repoInfo.CanonicalName, tag)
 	}
+	out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
```
Shouldn't this happen *before* the `"image manifest verified"` log? I just find the sequence a bit weird:

```
Image manifest verified for foo/bar:baz
Pulling from foo/bar
```

tiborvass replied:
@icecrime I read it as "It's pulling the image", not the "manifest". So it verifies the manifest first and then pulls the image data. Should we clarify it somehow?

jfrazelle replied:
The log will go to the Daemon logs anyways

On Wednesday, January 21, 2015, Tibor Vass <notifications@github.com> wrote:

> In graph/pull.go
> <https://github.com/docker/docker/pull/10235#discussion_r23285992>:
>
> >  	}
> > +	out.Write(sf.FormatStatus(tag, "Pulling from %s", repoInfo.CanonicalName))
>
> @icecrime <https://github.com/icecrime> I read it as "It's pulling the
> image", not the "manifest". So it verifies the manifest first and then
> pulls the image data. Should we clarify it somehow?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10235/files#r23285992>.
>

jfrazelle replied:
do you think this is ok or should I still move it?

tiborvass replied:
I'm personally okay with this, but what about the others?

tianon replied:
If it's just in the daemon log, that doesn't bother me. :)

(Nobody reads the daemon logs...)

icecrime replied:
Yes sorry I didn't notice this was going to the logs anyways. I'm okay with this.

icecrime commented:
Ping @dmp42 @dmcgowan: I think you guys should give your go on this.

dmcgowan commented on graph/pull.go:
```diff
@@ -491,7 +491,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 				io.Copy(tmpFile, utils.ProgressReader(ioutil.NopCloser(tarSumReader), int(l), out, sf, false, utils.TruncateID(img.ID), "Downloading"))
 
-				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum", nil))
+				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum - this is a tech preview, don't rely on it for security yet.", nil))
```
This will only flash for a fraction of a second in most cases

jfrazelle replied:
well I still think it needs to be said, do you suggest putting it somewhere else so it stays

dmcgowan replied:
I think the only other place it belongs is next to the verified message.  Keep in mind the current verified message will only show up for official images signed by official image build key.  While we might change the checksum method, the fact that we are verifying the checksum using tarsum IMO does not need to need the preview subtext.

jfrazelle replied:
I think it does need the preview subtext and I believe @shykes thought so
as well unless I misunderstood.

On Wed, Jan 21, 2015 at 3:13 PM, Derek McGowan <notifications@github.com>
wrote:

> In graph/pull.go
> <https://github.com/docker/docker/pull/10235#discussion_r23343298>:
>
> > @@ -491,7 +491,7 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
> >
> >  				io.Copy(tmpFile, utils.ProgressReader(ioutil.NopCloser(tarSumReader), int(l), out, sf, false, utils.TruncateID(img.ID), "Downloading"))
> >
> > -				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum", nil))
> > +				out.Write(sf.FormatProgress(utils.TruncateID(img.ID), "Verifying Checksum - this is a tech preview, don't rely on it for security yet.", nil))
>
> I think the only other place it belongs is next to the verified message.
> Keep in mind the current verified message will only show up for official
> images signed by official image build key. While we might change the
> checksum method, the fact that we are verifying the checksum using tarsum
> IMO does not need to need the preview subtext.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10235/files#r23343298>.
>

dmcgowan replied:
I may have misunderstood as well, that the subtext belongs with the "verification" claim not the checksum claim.

jfrazelle commented:
@dmcgowan @icecrime let me know about the latest changes

jfrazelle commented:
added a test

jessfraz committed:
Add test for pull verified

Docker-DCO-1.1-Signed-off-by: Jessica Frazelle <jess@docker.com> (github: jfrazelle)

dmcgowan commented:
I like it, LGTM

icecrime commented:
I like it too:

```
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
$ docker pull localhost:5000/icecrime/busybox:latest
latest: Pulling from localhost:5000/icecrime/busybox

511136ea3c5a: Pull complete
df7546f9f060: Pull complete
ea13149945cb: Pull complete
4986bf8c1536: Pull complete
localhost:5000/icecrime/busybox:latest: The image you are pulling has been verified.
This is a tech preview, don't rely on it for security yet.

Status: Downloaded newer image for localhost:5000/icecrime/busybox:latest
```

icecrime commented:
Once small nitpick: I'd replace the newline by a ` - ` to make it clear we're referring to the `"verified"` part, and not to the whole image pulling thing :p

Otherwise LGTM!

jfrazelle commented:
Updated!

tiborvass commented:
LGTM

fredlf commented on graph/pull.go:
```diff
@@ -553,6 +554,8 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
 
 	}
 
+	out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified - This is a tech preview, don't rely on it for security yet."))
```
Arnaud asked me to look at this and I'm just getting to it now. 
I would suggest the following: "The image you are pulling has been verified. *Important:* image verification is a tech preview feature and should not be relied on to provide security."

jfrazelle replied:
Ah ok that @fredlf I will ping you in a followup PR sorry we didn't wait.

On Wed, Jan 21, 2015 at 5:59 PM, Fred Lifton <notifications@github.com>
wrote:

> In graph/pull.go
> <https://github.com/docker/docker/pull/10235#discussion_r23350953>:
>
> > @@ -553,6 +554,8 @@ func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Wri
> >
> >  	}
> >
> > +	out.Write(sf.FormatStatus(repoInfo.CanonicalName+":"+tag, "The image you are pulling has been verified - This is a tech preview, don't rely on it for security yet."))
>
> Arnaud asked me to look at this and I'm just getting to it now.
> I would suggest the following: "The image you are pulling has been
> verified. *Important:* image verification is a tech preview feature and
> should not be relied on to provide security."
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10235/files#r23350953>.
>
